### PR TITLE
fix: report reputation changes correctly

### LINKED ISSUE #1038: Erroneous response is always reported as BadMessage
### Describe the bug

if a response is `Err` it is always returned as

https://github.com/paradigmxyz/reth/blob/c23a2215b0ddb7660c45b2f32bd422491f03c3c9/crates/net/network/src/fetch/mod.rs#L240-L246

which does not distinguish between timeouts, channel related errors

## TODO

* extend `EthResponseValidator` with function `fn reputation_change_err(&self) -> Option<ReputationChangeKind>` based on the error variant.
* used that function for determine the reputation to report here: https://github.com/paradigmxyz/reth/blob/c23a2215b0ddb7660c45b2f32bd422491f03c3c9/crates/net/network/src/fetch/mod.rs#L240-L246

### Steps to reproduce

-

### Node logs

_No response_

### Platform(s)

_No response_

### What version/commit are you on?

_No response_

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### PR DESCRIPTION
**Pull request by leruaa**
Closes #1038

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
this turned out to be a bit more complex than anticipated.

some errors are already handled when the session is dropped, or should be handled by the active session internally.

The only error worth handling here is actually `Timeout` imo just because we want to penalize slow peers during sync.

every other `Err` should be handled by the active session. And the downloader should only penalize `Ok` variants that are `BadMessages`.

This means, for this PR we need to keep the `is_err: bool` check to prevent followup requests, but should only penalize timeouts since this is only used during sync (at least atm).

I hope that made sense?

**Comment by gakonst** [comment]
Would love a test here that captures the edge case. We can't be changing stuff in P2P recklessly, as it works well-ish already and we don't want to introduce regressions.

**Review by mattsse — APPROVED** [review]
lgtm,
would appreciate a simple test for `on_block_headers_response` call with various error variants.

File: crates/net/network/src/fetch/mod.rs [hunk 1]
```diff
@@ -483,4 +483,34 @@ mod tests {
...
+            None
+        );
+        assert_eq!(
+            fetcher.on_block_headers_response(peer_id, Err(RequestError::ConnectionDropped)),
+            None
+        );
+        assert_eq!(
+            fetcher.on_block_headers_response(peer_id, Err(RequestError::UnsupportedCapability)),
+            None
+        );
```
Comment (gakonst) [inline crates/net/network/src/fetch/mod.rs:514]: Is this right? Wouldn't these return `BlockResponseOutcome::BadResponse` before, and now they return `None`? Seems weird?

**Review by gakonst — CHANGES_REQUESTED** [review]

File: crates/net/network/src/fetch/mod.rs (hunk 1 above)
Comment (leruaa) [reply crates/net/network/src/fetch/mod.rs:514]: As @mattsse [stated above](https://github.com/paradigmxyz/reth/pull/1086#pullrequestreview-1275310529), most of them are handled by the active session (Sorry, I've marked all his comments as solved).

Comment (mattsse) [reply crates/net/network/src/fetch/mod.rs:514]: this variant is actually not used here because we ensure a session has eth support.

we could remove this variant entirely. atm it would only be relevant if there's an incoming request for capability other than `eth`

**Review by mattsse — APPROVED** [review]
lgtm

File: crates/net/network/src/fetch/mod.rs (hunk 1 above)
Comment (mattsse) [reply crates/net/network/src/fetch/mod.rs:514]: I'll add some additional docs on reporting

**Comment by gakonst** [comment]
> we could remove this variant entirely. atm it would only be relevant if there's an incoming request for capability other than eth

let's keep it given we'll support more eth protocols
